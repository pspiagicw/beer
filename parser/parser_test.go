package parser

import (
	"testing"

	"github.com/gkampitakis/go-snaps/snaps"
	"github.com/kr/pretty"
	"github.com/pspiagicw/osy/ast"
	"github.com/pspiagicw/osy/lexer"
	"github.com/pspiagicw/osy/types"
	"github.com/stretchr/testify/assert"
)

func TestVariableAssignmentBoolean(t *testing.T) {
	input := `let status = false;`

	ast := &ast.Program{
		Statements: []ast.Statement{
			&ast.Assignment{
				Name: "status",
				Value: &ast.Constant{
					Value: "false",
					Type:  types.Boolean,
				},
			},
		},
	}

	testParser(t, input, ast)
}

func TestVariableAssignmentString(t *testing.T) {
	input := `let name = "pspiagicw";`

	ast := &ast.Program{
		Statements: []ast.Statement{
			&ast.Assignment{
				Name: "name",
				Value: &ast.Constant{
					Value: "pspiagicw",
					Type:  types.String,
				},
			},
		},
	}

	testParser(t, input, ast)
}

func TestVariableAssignmentInteger(t *testing.T) {
	input := `let a = 1;`

	ast := &ast.Program{
		Statements: []ast.Statement{
			&ast.Assignment{
				Name: "a",
				Value: &ast.Constant{
					Value: "1",
					Type:  types.Integer,
				},
			},
		},
	}

	testParser(t, input, ast)
}

func TestVariableDeclarationString(t *testing.T) {
	input := `def name string;`

	ast := &ast.Program{
		Statements: []ast.Statement{
			&ast.VariableDeclaration{
				Name: "name",
				Type: types.String,
			},
		},
	}

	testParser(t, input, ast)
}

func TestVariableDeclarationBoolean(t *testing.T) {
	input := `def status bool;`

	ast := &ast.Program{
		Statements: []ast.Statement{
			&ast.VariableDeclaration{
				Name: "status",
				Type: types.Boolean,
			},
		},
	}

	testParser(t, input, ast)
}

func TestVariableDeclarationInteger(t *testing.T) {
	input := `def a int;`

	ast := &ast.Program{
		Statements: []ast.Statement{
			&ast.VariableDeclaration{
				Name: "a",
				Type: types.Integer,
			},
		},
	}

	testParser(t, input, ast)
}

func TestReturnBoolean(t *testing.T) {
	input := `return true;`

	ast := &ast.Program{
		Statements: []ast.Statement{
			&ast.ReturnStatement{
				Value: &ast.Constant{
					Value: "true",
					Type:  types.Boolean,
				},
			},
		},
	}

	testParser(t, input, ast)
}

func TestReturnString(t *testing.T) {
	input := `return "hello";`

	ast := &ast.Program{
		Statements: []ast.Statement{
			&ast.ReturnStatement{
				Value: &ast.Constant{
					Value: "hello",
					Type:  types.String,
				},
			},
		},
	}

	testParser(t, input, ast)
}

func TestReturnFloat(t *testing.T) {
	input := `return 3.14;`

	ast := &ast.Program{
		Statements: []ast.Statement{
			&ast.ReturnStatement{
				Value: &ast.Constant{
					Value: "3.14",
					Type:  types.Float,
				},
			},
		},
	}

	testParser(t, input, ast)
}
func TestReturnInteger(t *testing.T) {
	input := `return 1;`

	ast := &ast.Program{
		Statements: []ast.Statement{
			&ast.ReturnStatement{
				Value: &ast.Constant{
					Value: "1",
					Type:  types.Integer,
				},
			},
		},
	}

	testParser(t, input, ast)
}

func testParser(t *testing.T, input string, expected ast.Node) {
	l := lexer.New(input)
	p := New(l)

	program := p.Parse()

	if p.HasError {
		for _, error := range p.Errors() {
			t.Errorf("PARSER ERROR: %v", error)
		}
		t.Fatalf("Parser Errors!")
	}

	if !assert.Equal(t, expected, program) {
		pretty.Println(program)
	}
	snaps.MatchSnapshot(t, program)
}
