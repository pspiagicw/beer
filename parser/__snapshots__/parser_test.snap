
[TestReturn - 1]
(*ast.Program)(nil)
---

[TestReturnInteger - 1]
&ast.Program{
    Statements: {
        &ast.ReturnStatement{
            Value: &ast.Constant{Value:"1", Type:0},
        },
    },
}
---

[TestReturnBoolean - 1]
&ast.Program{
    Statements: {
        &ast.ReturnStatement{
            Value: &ast.Constant{Value:"true", Type:3},
        },
    },
}
---

[TestFunctionDefinition - 1]
&ast.Program{
    Statements: {
    },
}
---

[TestVariableAssignmentBoolean - 1]
&ast.Program{
    Statements: {
        &ast.Assignment{
            Name:  "status",
            Value: &ast.Constant{Value:"false", Type:3},
        },
    },
}
---

[TestVariableAssignmentInteger - 1]
&ast.Program{
    Statements: {
        &ast.Assignment{
            Name:  "a",
            Value: &ast.Constant{Value:"1", Type:0},
        },
    },
}
---

[TestVariableDeclarationString - 1]
&ast.Program{
    Statements: {
        &ast.VariableDeclaration{Name:"name", Type:2},
    },
}
---

[TestVariableDeclarationBoolean - 1]
&ast.Program{
    Statements: {
        &ast.VariableDeclaration{Name:"status", Type:3},
    },
}
---

[TestVariableDeclarationInteger - 1]
&ast.Program{
    Statements: {
        &ast.VariableDeclaration{Name:"a", Type:0},
    },
}
---

[TestVariableAssignmentString - 1]
&ast.Program{
    Statements: {
        &ast.Assignment{
            Name:  "name",
            Value: &ast.Constant{Value:"pspiagicw", Type:2},
        },
    },
}
---

[TestReturnString - 1]
&ast.Program{
    Statements: {
        &ast.ReturnStatement{
            Value: &ast.Constant{Value:"hello", Type:2},
        },
    },
}
---

[TestReturnFloat - 1]
&ast.Program{
    Statements: {
        &ast.ReturnStatement{
            Value: &ast.Constant{Value:"3.14", Type:1},
        },
    },
}
---

[TestReturnAdditionAndSubtraction - 1]
&ast.Program{
    Statements: {
        &ast.ReturnStatement{
            Value: &ast.BinaryExpression{
                Right:    &ast.Constant{Value:"1", Type:0},
                Operator: 12,
                Left:     &ast.BinaryExpression{
                    Right:    &ast.Constant{Value:"3", Type:0},
                    Operator: 10,
                    Left:     &ast.Constant{Value:"2", Type:0},
                },
            },
        },
    },
}
---

[TestReturnAdd - 1]
&ast.Program{
    Statements: {
        &ast.ReturnStatement{
            Value: &ast.BinaryExpression{
                Right:    &ast.Constant{Value:"2", Type:0},
                Operator: 10,
                Left:     &ast.Constant{Value:"1", Type:0},
            },
        },
    },
}
---

[TestReturnSubstract - 1]
&ast.Program{
    Statements: {
        &ast.ReturnStatement{
            Value: &ast.BinaryExpression{
                Right:    &ast.Constant{Value:"2", Type:0},
                Operator: 12,
                Left:     &ast.Constant{Value:"1", Type:0},
            },
        },
    },
}
---

[TestReturnMultiply - 1]
&ast.Program{
    Statements: {
        &ast.ReturnStatement{
            Value: &ast.BinaryExpression{
                Right:    &ast.Constant{Value:"2", Type:0},
                Operator: 11,
                Left:     &ast.Constant{Value:"1", Type:0},
            },
        },
    },
}
---

[TestReturnDivide - 1]
&ast.Program{
    Statements: {
        &ast.ReturnStatement{
            Value: &ast.BinaryExpression{
                Right:    &ast.Constant{Value:"2", Type:0},
                Operator: 13,
                Left:     &ast.Constant{Value:"1", Type:0},
            },
        },
    },
}
---

[TestReturnMultiplicationAndDivisionPrecedence - 1]
&ast.Program{
    Statements: {
        &ast.ReturnStatement{
            Value: &ast.BinaryExpression{
                Right: &ast.BinaryExpression{
                    Right:    &ast.Constant{Value:"4", Type:0},
                    Operator: 13,
                    Left:     &ast.Constant{Value:"3", Type:0},
                },
                Operator: 10,
                Left:     &ast.Constant{Value:"2", Type:0},
            },
        },
    },
}
---

[TestReturnMultipleOperatorsPrecedence - 1]
&ast.Program{
    Statements: {
        &ast.ReturnStatement{
            Value: &ast.BinaryExpression{
                Right: &ast.BinaryExpression{
                    Right:    &ast.Constant{Value:"2", Type:0},
                    Operator: 13,
                    Left:     &ast.Constant{Value:"5", Type:0},
                },
                Operator: 12,
                Left:     &ast.BinaryExpression{
                    Right: &ast.BinaryExpression{
                        Right:    &ast.Constant{Value:"4", Type:0},
                        Operator: 11,
                        Left:     &ast.Constant{Value:"3", Type:0},
                    },
                    Operator: 10,
                    Left:     &ast.Constant{Value:"2", Type:0},
                },
            },
        },
    },
}
---
